src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+
+/* dt_accessor.h */
+
+int     tnt_dt_year         (dt_t dt);
+int     tnt_dt_month        (dt_t dt);
+int     tnt_dt_doy          (dt_t dt);
+int     tnt_dt_dom          (dt_t dt);
+
+/* dt_arithmetic.h definitions */
+
+typedef enum {
+    DT_EXCESS,
+    DT_LIMIT,
+    DT_SNAP
+} dt_adjust_t;
+
+dt_t   tnt_dt_add_years    (dt_t dt, int delta, dt_adjust_t adjust);
+dt_t   tnt_dt_add_quarters (dt_t dt, int delta, dt_adjust_t adjust);
+dt_t   tnt_dt_add_months   (dt_t dt, int delta, dt_adjust_t adjust);
+
+/* dt_parse_iso.h definitions */
+
+size_t tnt_dt_parse_iso_date (const char *str, size_t len, dt_t *dt);
+size_t tnt_dt_parse_iso_time (const char *str, size_t len, int *sod, int *nsec);
+size_t tnt_dt_parse_iso_zone_lenient(const char *str, size_t len, int *offset);
+
+/* Tarantool functions - datetime.c */
+
+int    datetime_to_string(char *buf, int len, const struct datetime * date);
+size_t datetime_strftime(char *buf, uint32_t len, const char *fmt,
+                        const struct datetime *date);
+void   datetime_now(struct datetime *now);
+
+]]
+
+local builtin = ffi.C
+local math_modf = math.modf
+local math_floor = math.floor
+
+local SECS_PER_DAY     = 86400
+local NANOS_PER_SEC    = 1000000000
+
+-- c-dt/dt_config.h
+
+-- Unix, January 1, 1970, Thursday
+local DT_EPOCH_1970_OFFSET = 719163
+
+
+local datetime_t = ffi.typeof('struct datetime')
+
+local function is_datetime(o)
+    return ffi.istype(datetime_t, o)
+end
+
+local function check_date(o, message)
+    if not is_datetime(o) then
+        return error(("%s: expected datetime, but received %s"):
+                     format(message, o), 2)
+    end
+end
+
+local function check_table(o, message)
+    if type(o) ~= 'table' then
+        return error(("%s: expected table %s"):format(message, o), 2)
+    end
+end
+
+local function check_str(s, message)
+    if not type(s) == 'string' then
+        return error(("%s: expected string, but received %s"):
+                     format(message, s), 2)
+    end
+end
+
+-- range may be of a form of pair {begin, end} or
+-- tuple {begin, end, negative}
+-- negative is a special value (so far) used for days only
+local function check_range(v, range, txt)
+    local len = #range
+    assert(len == 2 or len == 3)
+
+    local left, right, neg = unpack(range)
+    if neg == v or (v >= left and v <= right) then
+        return
+    end
+
+    if neg == nil then
+        error(('value %d of %s is out of allowed range [%d, %d]'):
+              format(v, txt, left, right), 2)
+    else
+        error(('value %d of %s is out of allowed range [%d, %d..%d]'):
+              format(v, txt, neg, left, right), 2)
+    end
+end
+
+local function nyi(msg)
+    local text = 'Not yet implemented'
+    if msg ~= nil then
+        text = ("%s : '%s'"):format(text, msg)
+    end
+    error(text, 3)
+end
+
+-- offset if __seconds__ of 1970-01-01 from 0000-01-01
+local SECS_EPOCH_OFFSET = (DT_EPOCH_1970_OFFSET * SECS_PER_DAY)
+
```

Comment:
You don't need `()` around this expression.

Reply:
Done